### PR TITLE
Signum optimizer

### PR DESCRIPTION
**Pull request by yuxiangw**
## Description ##
Added the C++ implementation of the Signum optimizer. 

Bernstein, Wang, Azizzadenesheli and Anandkumar (2017) "The Signum optimiser: a theory of momentum in quantised stochastic optimisation" 
[Link to pdf](https://jeremybernste.in/projects/amazon/signum.pdf)

What's also included is the implementation of an option to do the alternative version of weight decay regularization due to Loshchilov and Hutter via option 'wd_lh'.
"Fixing Weight Decay Regularization in Adam"
[Link to arxiv](https://arxiv.org/abs/1711.05101)

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Added the Signum optimizer to mxnet's list of optimizers 
- [x] A special case is SignSGD optimizer, a stand-alone implementation whenever "momentum" is set to 0.

## Comments ##
- TODO1: add sparse matrix support for this optimizer
- TODO2: Take advantage of the 1-bit gradient compression interpretation of SignSGD and Signum.
- TODO3: Adding 'wd_lh' support for Adam and other adaptive gradient optimizers.

### CODE REVIEW

File: python/mxnet/optimizer.py [hunk 1]
```diff
@@ -529,6 +534,69 @@ def update_multi_precision(self, index, weight, grad, state):
         self._update_impl(index, weight, grad, state,
                           multi_precision=use_multi_precision)
 
+
+
+@register
+class Signum(Optimizer):
+    """The SGD optimizer with momentum and weight decay.
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer.py]: The one line summary should also mention it only takes the sign. Otherwise the readers don't know it until they see line 547

File: src/operator/optimizer_op-inl.h [hunk 2]
```diff
@@ -1088,6 +1090,146 @@ inline void FtrlUpdateEx(const nnvm::NodeAttrs& attrs,
...
 }
 
+
+// Implementation for signSGD and Signum
+
+struct SignSGDParam : public dmlc::Parameter<SignSGDParam> {
+  float lr;
+  float wd;
+  float rescale_grad;
+  float clip_gradient;
```
Comment (eric-haibin-lin) [inline src/operator/optimizer_op-inl.h]: If the `clip_gradient ` param has no effect on both SignSGD and Signum, can we just remove this param from `signsgd_update` and `signum_update`?  That would also simply the c++ kernels

File: python/mxnet/optimizer.py [hunk 3]
```diff
@@ -529,6 +534,69 @@ def update_multi_precision(self, index, weight, grad, state):
...
+
+    This optimizer accepts the following parameters in addition to those accepted
+    by :class:`.Optimizer`.
+
+    Parameters
+    ----------
+    momentum : float, optional
+       The momentum value.
+    wd_lh : float, optitional
+       The amount of decoupled weight decay regularization.
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer.py]: Let's also add a reference/link to the original paper

File: python/mxnet/optimizer.py [hunk 4]
```diff
@@ -57,6 +58,10 @@ class Optimizer(object):
         The weight decay (or L2 regularization) coefficient. Modifies objective
         by adding a penalty for having large weights.
 
+    wd_lh: float, optional
```
[the remark this answers is not in the document]
Comment (yuxiangw) [reply python/mxnet/optimizer.py]: I added that to the constructor at some point, cuz wd_lh is something more generally applicable to other algorithms too (in particular, Adam).

File: src/operator/optimizer_op-inl.h (hunk 2 above)
Comment (yuxiangw) [reply src/operator/optimizer_op-inl.h]: It has an effect on Signum. Because it will lead to different result whether we use gradient or clipped gradient for calculating momentum.

Comment (eric-haibin-lin) [reply src/operator/optimizer_op-inl.h]: Ah, I see. Thanks for the explanation!

File: python/mxnet/optimizer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yuxiangw) [reply python/mxnet/optimizer.py]: removed that line.

File: python/mxnet/optimizer.py (hunk 1 above)
Comment (yuxiangw) [reply python/mxnet/optimizer.py]: added details to the doc accordingly.

File: python/mxnet/optimizer.py (hunk 3 above)
Comment (yuxiangw) [reply python/mxnet/optimizer.py]: added the temp link to pdf hosted on jeremy's site. will update to arxiv or a published version when they are ready.

File: src/operator/optimizer_op.cc [hunk 5]
```diff
@@ -35,6 +35,74 @@ DMLC_REGISTER_PARAMETER(AdamParam);
 DMLC_REGISTER_PARAMETER(RMSPropParam);
 DMLC_REGISTER_PARAMETER(RMSPropAlexParam);
 DMLC_REGISTER_PARAMETER(FtrlParam);
+DMLC_REGISTER_PARAMETER(SignSGDParam);
+DMLC_REGISTER_PARAMETER(SignumParam);
+
+NNVM_REGISTER_OP(signsgd_update)
+// MXNET_ADD_SPARSE_OP_ALIAS(signsgd_update)
+.describe(R"code(Update function for SignSGDoptimizer.
```
[the remark this answers is not in the document]
Comment (yuxiangw) [reply src/operator/optimizer_op.cc]: done. and added the math description block similar to other optimizers.

File: src/operator/optimizer_op.cc [hunk 6]
```diff
@@ -35,6 +35,74 @@ DMLC_REGISTER_PARAMETER(AdamParam);
...
+It updates the weights using::
+
+ weight = weight - learning_rate * sign(gradient)
+
+
+** Sparse matrix not supported for this optimizer yet.
```
[the remark this answers is not in the document]
Comment (yuxiangw) [reply src/operator/optimizer_op.cc]: done.

File: src/operator/optimizer_op.cc [hunk 7]
```diff
@@ -35,6 +35,74 @@ DMLC_REGISTER_PARAMETER(AdamParam);
...
+
+NNVM_REGISTER_OP(signum_update)
+// MXNET_ADD_SPARSE_OP_ALIAS(signum_update)
+.describe(R"code(SIGN momentUM (Signum) optimizer.
+
+ weight = weight - learning_rate * sign(momentum)
+
+Where the parameter ``momentum`` is the decay rate of momentum estimates at each epoch.
+
+** Sparse matrix not supported for this optimizer yet.
```
[the remark this answers is not in the document]
Comment (yuxiangw) [reply src/operator/optimizer_op.cc]: Done.

File: src/operator/optimizer_op.cc [hunk 8]
```diff
@@ -35,6 +35,74 @@ DMLC_REGISTER_PARAMETER(AdamParam);
...
+
+If weight and momentum are both of ``row_sparse`` storage type,
```
[the remark this answers is not in the document]
Comment (yuxiangw) [reply src/operator/optimizer_op.cc]: Done.

File: src/operator/optimizer_op.cc [hunk 9]
```diff
@@ -35,6 +35,74 @@ DMLC_REGISTER_PARAMETER(AdamParam);
...
+      v[row] = momentum[row] * v[row] - learning_rate * gradient[row]
+      weight[row] += v[row]
+
+)code" ADD_FILELINE)
+.set_num_inputs(3)
+.set_num_outputs(1)
+.set_attr_parser(ParamParser<SignumParam>)
+.set_attr<nnvm::FInferShape>("FInferShape", ElemwiseShape<3, 1>)
+.set_attr<nnvm::FInferType>("FInferType", ElemwiseType<3, 1>)
+.set_attr<FInferStorageType>("FInferStorageType", ElemwiseStorageType<3, 1, false, true, false>)
```
[the remark this answers is not in the document]
Comment (yuxiangw) [reply src/operator/optimizer_op.cc]: Done.

File: src/operator/optimizer_op.cc [hunk 10]
```diff
@@ -35,6 +35,74 @@ DMLC_REGISTER_PARAMETER(AdamParam);
...
+.set_attr<nnvm::FMutateInputs>("FMutateInputs",
+  [](const nnvm::NodeAttrs& attrs) {
+    return std::vector<uint32_t>{2};
+  })
+.set_attr<FCompute>("FCompute<cpu>", SignumUpdate<cpu>)
+// .set_attr<FComputeEx>("FComputeEx<cpu>", SGDMomUpdateEx<cpu>)
```
[the remark this answers is not in the document]
Comment (yuxiangw) [reply src/operator/optimizer_op.cc]: Done.

File: src/operator/optimizer_op.cu [hunk 11]
```diff
@@ -28,6 +28,14 @@
 namespace mxnet {
 namespace op {
 
+NNVM_REGISTER_OP(signsgd_update)
+.set_attr<FCompute>("FCompute<gpu>", SignSGDUpdate<gpu>);
+// .set_attr<FComputeEx>("FComputeEx<gpu>", SignSGDUpdateEx<gpu>);
```
[the remark this answers is not in the document]
Comment (yuxiangw) [reply src/operator/optimizer_op.cu]: Done.

**Comment by eric-haibin-lin** [comment]
There are new conflicts now. Do you mind resolving them again?
BTW - the files under cpp-package are only needed if you use cpp as front end to train networks. Do you actually need it?

**Comment by yuxiangw** [comment]
Done fixing the conflicts.

**Comment by eric-haibin-lin** [comment]
@lx75249 could you help review the code for cpp-package?

**Comment by lx75249** [comment]
@eric-haibin-lin LGTM

File: python/mxnet/optimizer.py [hunk 12]
```diff
@@ -534,6 +535,66 @@ def update_multi_precision(self, index, weight, grad, state):
...
+        if self.clip_gradient:
+            kwargs['clip_gradient'] = self.clip_gradient
+        if self.wd_lh:
+            kwargs['wd_lh'] = self.wd_lh
+
+        if state is not None:
+            signum_update(weight, grad, state, out=weight,
+                          lr=lr, wd=wd, **kwargs)
+        else:
+            signsgd_update(weight, grad, out=weight,
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: what's this?

File: python/mxnet/optimizer.py [hunk 13]
```diff
@@ -534,6 +535,66 @@ def update_multi_precision(self, index, weight, grad, state):
...
 
+@register
+class Signum(Optimizer):
+    """The Signum optimizer that takes the sign of gradient or momentum.
+
+    The optimizer updates the weight by:
+
+        rescaled_grad = rescale_grad * clip(grad, clip_gradient) + wd * weight
+        state = momentum * state + (1-momentum)*rescaled_grad
+        weight = (1 - lr * wd_lh) * weight - lr * sign(state)
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: what's wd_lh? Is it from the original paper?

File: python/mxnet/optimizer.py [hunk 14]
```diff
@@ -534,6 +535,66 @@ def update_multi_precision(self, index, weight, grad, state):
...
+        kwargs = {'rescale_grad': self.rescale_grad}
+        if self.momentum > 0:
+            kwargs['momentum'] = self.momentum
+        if self.clip_gradient:
+            kwargs['clip_gradient'] = self.clip_gradient
+        if self.wd_lh:
+            kwargs['wd_lh'] = self.wd_lh
+
+        if state is not None:
+            signum_update(weight, grad, state, out=weight,
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: call these signum_momentum_update and signum_update to be consistent with others

File: python/mxnet/optimizer.py (hunk 13 above)
Comment (yuxiangw) [reply python/mxnet/optimizer.py]: It is an alternative weight decay. See the descriptions.

File: python/mxnet/optimizer.py (hunk 12 above)
Comment (yuxiangw) [reply python/mxnet/optimizer.py]: well, signsgd takes the sign of stochastic gradient.

File: python/mxnet/optimizer.py (hunk 14 above)
Comment (yuxiangw) [reply python/mxnet/optimizer.py]: RE: naming.  
- signum means SIGN momentUM. So the semantics of the momentum is already in there. -
- SignSGD is the special case of Signum that goes without momentum. And it has been used before.

Unless we change the names in our paper, let's keep them the way they are.

File: python/mxnet/optimizer.py (hunk 13 above)
Comment (eric-haibin-lin) [reply python/mxnet/optimizer.py]: Since `wd_lh` is new, I suggest put a reference link to the original paper by Loshchilov and Frank Hutter in the documentation

**Comment by yuxiangw** [comment]
Added the reference the documentation as suggested. Thanks guys for reviewing the PR!

**Comment by piiswrong** [comment]
Thanks
